src/main/java/com/volmit/iris/core/TreeManager.java

```diff
@@ -0,0 +1,217 @@
+    }
+
+    /**
+     * Finds a single object placement (which may contain more than one object) for the requirements species, location & size
+     * @param worldAccess The world to access (check for biome, region, dimension, etc)
+     * @param location The location of the growth event (For biome/region finding)
+     * @param type The bukkit TreeType to match
+     * @param size The size of the sapling area
+     * @return An object placement which contains the matched tree, or null if none were found / it's disabled.
+     */
+    private IrisObjectPlacement getObjectPlacement(IrisAccess worldAccess, Location location, IrisTreeType type, IrisTreeSize size) {
+
+        KList<IrisObjectPlacement> placements = new KList<>();
+
+        // Retrieve objectPlacements of type `species` from biome
+        IrisBiome biome = worldAccess.getBiome(location.getBlockX(), location.getBlockY(), location.getBlockZ());
+        placements.addAll(matchObjectPlacements(biome.getObjects(), size, type));
+
+        // Add more or find any in the region
+        if (worldAccess.getCompound().getRootDimension().getTreeSettings().getMode().equals(IrisTreeModes.ALL) || placements.isEmpty()){
+            IrisRegion region = worldAccess.getCompound().getDefaultEngine().getRegion(location.getBlockX(), location.getBlockZ());
+            placements.addAll(matchObjectPlacements(region.getObjects(), size, type));
+        }
+
+        // Add more or find any in the dimension
+        if (worldAccess.getCompound().getRootDimension().getTreeSettings().getMode().equals(IrisTreeModes.ALL) || placements.isEmpty()){
+            placements.addAll(matchObjectPlacements(worldAccess.getCompound().getRootDimension().getObjects(), size, type));
+        }
+
+        // Check if no matches were found, return a random one if they are
+        return placements.isNotEmpty() ? placements.getRandom(RNG.r) : null;
+    }
+
+    /**
+     * Filters out mismatches and returns matches
+     * @param objects The object placements to check
+     * @param size The size of the sapling area to filter with
+     * @param type The type of the tree to filter with
+     * @return A list of objectPlacements that matched. May be empty.
+     */
+    private KList<IrisObjectPlacement> matchObjectPlacements(KList<IrisObjectPlacement> objects, IrisTreeSize size, IrisTreeType type) {
+        KList<IrisObjectPlacement> objectPlacements = new KList<>();
+        objects.stream()
+                .filter(objectPlacement -> objectPlacement.getTreeOptions().isEnabled())
+                .filter(objectPlacement -> objectPlacement.getTreeOptions().getTrees().stream().anyMatch(irisTree ->
+                        irisTree.getSizes().stream().anyMatch(treeSize -> treeSize == IrisTreeSize.ANY || treeSize == size) &&
+                        irisTree.getTreeTypes().stream().anyMatch(treeType -> treeType == IrisTreeType.ANY || treeType == type)))
+                .forEach(objectPlacements::add);
```

Comment:
:P Could to collect() then just return the whole thing.

Reply:
Updated to:
```java
    private KList<IrisObjectPlacement> matchObjectPlacements(KList<IrisObjectPlacement> objects, IrisTreeSize size, TreeType type) {

        Predicate<IrisTree> isValid = irisTree -> (
                irisTree.isAnySize() || irisTree.getSizes().stream().anyMatch(treeSize -> treeSize == IrisTreeSize.ANY || treeSize == size)) && (
                irisTree.isAnyTree() || irisTree.getTreeTypes().stream().anyMatch(treeType -> treeType == type));

        objects.removeIf(objectPlacement -> objectPlacement.getTrees().stream().noneMatch(isValid));

        return objects;
    }```

Reply:
(Still need to implement the sizes properly though. This is just an in between step to ensure I did everything right thus far)